src/NonCumulative/Soundness/Properties/LogRel.agda

```diff
@@ -86,4 +98,712 @@ Glu-wellfounded-≡-GluΛ {j} {k} = implicit-extensionality fext (fext (λ l<k 
+    { UT = UT 
+    ; ⊢UT = ⊢UT 
+    ; T≈ = T≈ 
+    ; krip = λ ⊢σ → helper kA kA′ (krip ⊢σ) 
+    }
+    where helper : (kA : A ≈ A′ ∈ PERDef.𝕌 k (λ l<k → 𝕌-wellfounded (j + k) (Li≤ refl l<k))) → 
+                  (kA′ : A′ ≈ A ∈ PERDef.𝕌 k (λ l<k → 𝕌-wellfounded (j + k) (Li≤ refl l<k))) → 
+                  Glu.⟦ k , Glu-wellfounded k , (λ {l} l<k → 𝕌-wellfounded (j + k) (Li≤ refl l<k))⟧ Δ ⊢ sub UT σ ® kA → 
+                  -----------------------------------
+                  Glu.⟦ k , Glu-wellfounded k , (λ {l} l<k → 𝕌-wellfounded (j + k) (Li≤ refl l<k))⟧ Δ ⊢ sub UT σ ® kA′
+          helper kA kA′ T® rewrite 𝕌-wf-gen {j + k} k (λ l<k → Li≤ refl l<k) = ®-swap kA kA′ T®
+    
+  ®El-swap : ∀ {i} (A≈B : A ≈ B ∈ 𝕌 i)
+             (B≈A : B ≈ A ∈ 𝕌 i) →
+             Γ ⊢ t ∶ T ®[ i ] a ∈El A≈B →
+             ----------------------------
+             Γ ⊢ t ∶ T ®[ i ] a ∈El B≈A
+  ®El-swap {_} {_} {Γ} {t} {T} {i = i} (ne′ C≈C′) (ne C′≈C _ _) (ne c≈c refl _ , glu) = (ne c≈c refl refl) , record 
+    { t∶T = t∶T
+    ; ⊢T = ⊢T
+    ; krip = λ ⊢σ → krip′ ⊢σ 
+    }
+    where
+      open GluNe glu
+      krip′ : Δ ⊢w σ ∶ Γ →
+               -----------------------------------
+               Δ ⊢ sub T σ ≈ Ne⇒Exp (proj₁ (C′≈C (L.length Δ))) ∶[ ℕ.suc i ] Se i
+                 × Δ ⊢ sub t σ ≈ Ne⇒Exp (proj₁ (c≈c (L.length Δ))) ∶[ i ] sub T σ
+      krip′ {Δ} {σ} ⊢σ
+        with C≈C′ (len Δ) | C′≈C (len Δ) | krip ⊢σ
+      ...  | _ , ↘u , _ | _ , _ , ↘u₁ | Tσ≈ , tσ≈ 
+           rewrite Re-det ↘u ↘u₁ = Tσ≈ , tσ≈
+  ®El-swap N′ N′ T® = T®
+  ®El-swap U′ (U i≡1+j j≡j′) T® 
+    rewrite ≡-irrelevant i≡1+j refl = T®
+  ®El-swap (Π′ {j} {k} jA RT) (Π i≡maxjk jA′ RT′ j≡j′ k≡k′) record { t∶T = t∶T ; a∈El = a∈El ; IT = IT ; OT = OT ; ⊢IT = ⊢IT ; ⊢OT = ⊢OT ; T≈ = T≈ ; krip = krip } 
+    rewrite ≡-irrelevant i≡maxjk refl | Glu-wellfounded-≡-GluΛI {j} {k} | Glu-wellfounded-≡-GluΛO {j} {k} = record
+    { t∶T = t∶T
+    ; a∈El = El-sym (Π′ jA RT) (Π′ jA′ RT′) a∈El
+    ; IT = IT
+    ; OT = OT
+    ; ⊢IT = ⊢IT
+    ; ⊢OT = ⊢OT
+    ; T≈ = T≈
+    ; krip = λ ⊢σ → let open ΛRel (krip ⊢σ) in record
+      { IT-rel = IT-helper jA jA′ IT-rel 
+      ; ap-rel = λ s® b∈ → ap-helper refl jA jA′ RT RT′ b∈ s® ap-rel 
+      } 
+    }
+    where IT-helper : ∀ {k′} → 
+                      (jA : A ≈ A′ ∈ PERDef.𝕌 j (λ l<j → 𝕌-wellfounded (max j k′) (ΠI≤ refl l<j))) → 
+                      (jA′ : A′ ≈ A ∈ PERDef.𝕌 j (λ l<j → 𝕌-wellfounded (max j k′) (ΠI≤ refl l<j))) → 
+                      Glu.⟦ j , Glu-wellfounded j , (λ l<j → 𝕌-wellfounded (max j k′) (ΠI≤ refl l<j))⟧ Δ ⊢ sub IT σ ® jA → 
+                      -----------------------------------
+                      Glu.⟦ j , Glu-wellfounded j , (λ l<j → 𝕌-wellfounded (max j k′) (ΠI≤ refl l<j))⟧ Δ ⊢ sub IT σ ® jA′
+          IT-helper {k′ = k′} jA jA′ T® rewrite 𝕌-wf-gen {max j k′} j (λ l<j → ΠI≤ refl l<j) = ®-swap jA jA′ T®
+
+          ap-helper : ∀ {k′} → 
```

Comment:
this is a bit too much, making the helper very challenging to read. let's discuss it later